docs/writing-rules/data-flow/taint-mode.md

```diff
@@ -31,90 +31,116 @@ Here Semgrep tracks the data returned by `get_user_input()`, which is the source
+| `exact`           | {`false`, `true`}         | `false` | See [_Exact sources_](#exact-sources).                                 |
+| `by-side-effect`  | {`false`, `true`, `only`} | `false` | See [_Sources by side-effect_](#sources-by-side-effect).               |
+| `control` (Pro) 🧪 | {`false`, `true`}         | `false` | See [_Control sources_](#control-sources-pro-).                           |
 
-:::note
-The taint engine does track taint **per variable** and not **per object in memory**. The taint engine does not perform alias analysis at present.
-:::
+Example:
 
-Sources
--------
+```yaml
+pattern-sources:
+- pattern: source(...)
+```
 
-A taint source is specified by a pattern. **Any** subexpression that this pattern matches is regarded as a source of taint.
+### Exact sources
 
-:::note
 Given the source specification below, and a piece of code such as `source(sink(x))`, the call `sink(x)` is reported as a tainted sink.
-The reason is that the pattern `source(...)` matches all of `source(sink(x))`, and that makes Semgrep label every subexpression in that piece of code as being a source. In particular, `x` is a source, and it is being passed into `sink`!
 
 ```yaml
 pattern-sources:
 - pattern: source(...)
 ```
-:::
 
-### Function arguments as sources
+The reason is that the pattern `source(...)` matches all of `source(sink(x))`, and that makes Semgrep consider every subexpression in that piece of code as being a source. In particular, `x` is a source, and it is being passed into `sink`!
 
-To specify that an argument of a function must be considered a taint source, simply write a pattern that matches that argument:
+<iframe src="https://semgrep.dev/embed/editor?snippet=eqYN8" border="0" frameBorder="0" width="100%" height="432"></iframe>
 
-```javascript
+This is the default for historical reasons, but it may change in the future.
+
+It is possible to instruct Semgrep to only consider as taint sources the "exact" matches of a souce pattern by setting `exact: true`:
+
+```yaml
 pattern-sources:
-  - patterns:
-    - pattern-inside: function ($REQ, ...) {...}
-    - focus-metavariable: $REQ
+- pattern: source(...)
+  exact: true
 ```
 
+Once the source is "exact", Semgrep will no longer consider subexpressions as taint sources, and `sink(x)` inside `source(sink(x))` will not be reported as a tainted sink (unless `x` is tainted in some other way).
+
+<iframe src="https://semgrep.dev/embed/editor?snippet=Zq5ow" border="0" frameBorder="0" width="100%" height="432"></iframe>
+
+For many rules this distinction is not very meaningful because it does not always make sense that a sink occurs inside the arguments of a source function.
+
+:::note
+If one of your rules relies on non-exact matching of sources, we advice you to make it explicit with `exact: false`, even if it is the current default, so that your rule does not break if the default changes.
+:::
+
 ### Sources by side-effect
```

Comment:
I'm just trying to wrap my head around the idea of a tainted side-effect. Given:
```
function add10(x) {
x = x + 10
return x
}
```
I assume the above is a pure function (no side effects), deterministic output.

So a side-effect would be the value of y in the ff example
```
function addy(y, x) {
y = getUserInputNumber()
x = x + y
return x
```
So if we wanted to catch `y`, we'd say:
```
 - patterns:
      - pattern: addy($Y, $X)
      - focus-metavariable: $Y
    by-side-effect: true
```
Is this correct?

Reply:
The idea seems correct although the example code not be the best one. If you change `addy(y, x)` to `addy(ys, x)`, `y = getUserInputNumber()` to `ys.add(getUserInputNumber())`, and `x = x + y` to `x = x + ys.pop()`, then yeah it's clearly correct. I'm no Python expert but I'm just stressing that `y` would be a mutable object that is being mutated inside `addy`. Numbers in principle are not mutable in Python, I think, and if you `y = getUserInputNumber()` then you're just replacing the value of `y` without altering what has been passed as argument.

Reply:
Got it. Yeah, it's hard to think of examples (at least for me) sometimes.